openssl/src/ts.rs

```diff
@@ -0,0 +1,457 @@
+    }
+
+    /// Sets whether to request the public key certificate in the response.
+    ///
+    /// This corresponds to `TS_REQ_set_cert_req`.
+    pub fn set_cert_req(&mut self, cert_req: bool) -> Result<(), ErrorStack> {
+        unsafe { cvt(ffi::TS_REQ_set_cert_req(self.0.as_ptr(), cert_req as c_int)).map(|_| ()) }
+    }
+
+    /// Consume the builder and return the timestamp request.
+    pub fn build(self) -> TsReq {
+        self.0
+    }
+}
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_RESP;
+    fn drop = ffi::TS_RESP_free;
+
+    /// A time-stamping response.
+    pub struct TsResp;
+
+    /// Reference to `TsResp`.
+    pub struct TsRespRef;
+}
+
+impl TsResp {
+    from_der! {
+        /// Deserializes a DER-encoded TimeStampResp structure.
+        ///
+        /// This corresponds to [`d2i_TS_RESP`].
+        ///
+        /// [`d2i_TS_RESP`]: https://www.openssl.org/docs/man1.1.0/man3/d2i_TS_RESP.html
+        from_der,
+        TsResp,
+        ffi::d2i_TS_RESP
+    }
+}
+
+impl TsRespRef {
+    to_der! {
+        /// Serializes the timestamp request into a DER-encoded TimeStampResp structure.
+        ///
+        /// This corresponds to [`i2d_TS_RESP`].
+        ///
+        /// [`i2d_TS_RESP`]: https://www.openssl.org/docs/man1.1.0/man3/i2d_TS_RESP.html
+        to_der,
+        ffi::i2d_TS_RESP
+    }
+
+    /// Verifies a timestamp response.
+    ///
+    /// This corresponds to `TS_RESP_verify_response`.
+    pub fn verify(&self, context: &TsVerifyContext) -> Result<(), ErrorStack> {
+        unsafe {
+            cvt(ffi::TS_RESP_verify_response(
+                context.as_ptr(),
+                self.as_ptr(),
+            ))
+            .map(|_| ())
+        }
+    }
+}
+
+bitflags! {
+    /// Flags controlling timestamp verification behaviour.
+    pub struct VerifyFlags: c_uint {
+        const SIGNATURE = ffi::TS_VFY_SIGNATURE;
+        const VERSION = ffi::TS_VFY_VERSION;
+        const POLICY = ffi::TS_VFY_POLICY;
+        const IMPRINT = ffi::TS_VFY_IMPRINT;
+        const DATA = ffi::TS_VFY_DATA;
+        const NONCE = ffi::TS_VFY_NONCE;
+        const SIGNER = ffi::TS_VFY_SIGNER;
+        const TSA_NAME = ffi::TS_VFY_TSA_NAME;
+
+        const ALL_IMPRINT = ffi::TS_VFY_ALL_IMPRINT;
+        const ALL_DATA = ffi::TS_VFY_ALL_DATA;
+    }
+}
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_VERIFY_CTX;
+    fn drop = ffi::TS_VERIFY_CTX_free;
+
+    /// A context object specifying time-stamping response verification parameters.
+    pub struct TsVerifyContext;
+
+    /// Reference to `TsVerifyContext`.
+    pub struct TsVerifyContextRef;
+}
+
+impl TsVerifyContext {
+    /// Construct a verify context from a timestamping request.
+    ///
+    /// Corresponds to `TS_REQ_to_TS_VERIFY_CTX`.
+    pub fn from_req(request: &TsReqRef) -> Result<TsVerifyContext, ErrorStack> {
+        unsafe {
+            ffi::init();
```

Comment:
This isn't necessary since a TsReqRef already exists.

Reply:
Removed, thanks!